loopy/target/execution.py

```diff
@@ -827,15 +827,15 @@ def get_typed_and_scheduled_translation_unit_uncached(
             # FIXME: This is not so nice. This transfers types from the
             # subarrays of sep-tagged arrays to the 'main' array, because
             # type inference fails otherwise.
-            with arg_to_dtype.mutate() as mm:
-                for name, sep_info in self.sep_info.items():
-                    if entry_knl.arg_dict[name].dtype is None:
-                        for sep_name in sep_info.subarray_names.values():
-                            if sep_name in arg_to_dtype:
-                                mm.set(name, arg_to_dtype[sep_name])
-                                del mm[sep_name]
+            mm = dict(arg_to_dtype)
+            for name, sep_info in self.sep_info.items():
+                if entry_knl.arg_dict[name].dtype is None:
+                    for sep_name in sep_info.subarray_names.values():
+                        if sep_name in arg_to_dtype:
+                            mm[name] = arg_to_dtype[sep_name]
+                            del mm[sep_name]
 
-                arg_to_dtype = mm.finish()
+            arg_to_dtype = constantdict(mm)
```

Comment:
Could this remain on the mutation interface?

Reply:
Changed in [ba93a2f](https://github.com/inducer/loopy/pull/884/commits/ba93a2f101b0ae665a2ab425d7fc43b26c1bf5ae)